File: 06 - ANNs I/mlp/mlp.go
package mlp

import (
	"math"
	"math/rand"

	"gonum.org/v1/gonum/mat"
)

type Mlp struct {
	NumInputs    int
	HiddenLayers []int
	NumOutputs   int
	weights      []mat.Matrix
}

func apply(fn func(i, j int, v float64) float64, m mat.Matrix) mat.Matrix {
	r, c := m.Dims()
	o := mat.NewDense(r, c, nil)
	o.Apply(fn, m)
	return o
}

func sigmoid(r, c int, z float64) float64 {
	return 1.0 / (1 + math.Exp(-1*z))
}

func generateRandom(n, m int) mat.Matrix {
	data := make([]float64, n*m)
	for i := range data {
		data[i] = rand.NormFloat64()
	}
	return mat.NewDense(n, m, data)
}

func (m *Mlp) Init() {
	layers := make([]int, 0)

	layers = append(layers, m.NumInputs)
	layers = append(layers, m.HiddenLayers...)
	layers = append(layers, m.NumOutputs)

	for i := 0; i < len(layers)-1; i++ {
		m.weights = append(m.weights, generateRandom(layers[i], layers[i+1]))
	}
}

func (m *Mlp) Activate(inputs mat.Matrix) mat.Matrix {
	activation := inputs

	for i := 0; i < len(m.weights); i++ {
		r, _ := activation.Dims()
		_, c := m.weights[i].Dims()
		dotProduct := mat.NewDense(r, c, nil)
		dotProduct.Product(activation, m.weights[i])
		activation = apply(sigmoid, dotProduct).(*mat.Dense)
	}

	return activation
}

File: 08 - ANNs III/python/ml/mlp.py
import numpy as np


class MLP(object):
    """A Multilayer Perceptron class.
    """

    def __init__(self, numInputs=3, hiddenLayers=(3, 3), numOutputs=1):
        """Constructor for the MLP. Takes the number of inputs,
            a variable number of hidden layers, and number of outputs

        Args:
            numInputs (int): Number of inputs
            hiddenLayers (list): A list of ints for the hidden layers
            numOutputs (int): Number of outputs
        """

        self.numInputs = numInputs
        self.hiddenLayers = hiddenLayers
        self.numOutputs = numOutputs

        # create a generic representation of the layers
        layers = [numInputs] + list(hiddenLayers) + [numOutputs]

        # create random connections for the layers
        weights = []
        for i in range(len(layers)-1):
            w = np.random.rand(layers[i], layers[i+1])
            weights.append(w)
        self.weights = weights

        # save activations/errors per layer
        activations = []
        errors = []
        for i in range(len(layers)):
            a = np.zeros(layers[i])
            e = a[:]
            activations.append(a)
            errors.append(e)
        self.activations = activations
        self.errors = errors

        # save derivatives per layer
        derivatives = []
        for i in range(len(layers) - 1):
            d = np.zeros(layers[i+1])
            derivatives.append(d)
        self.derivatives = derivatives

    def activate(self, input):
        """Computes activation of the network based on input signals.

        Args:
            input (list): Input signals

        Returns:
            output (list): Output values
        """

        # the input layer activation is just the input itself
        activation = input
        # save the activation for backpropogation
        self.activations[0][:] = activation

        # iterate through the network layers
        for i in range(len(self.weights)):
            # calculate dot product between previous activation and weights
            dotProduct = np.dot(activation, self.weights[i])

            # apply sigmoid activation function
            activation = self._sigmoid(dotProduct)

            # save the activation for backpropogation
            self.activations[i+1][:] = activation

        # return output layer activation
        return activation[0, :]

    def backActivate(self, error):
        """Backpropogates an error signal.

        Args:
            error (ndarray): The error to backprop.

        Returns:
            error (float): The final error of the input
        """

        # save the error
        self.errors[-1][:] = error

        # iterate backwards through the network layers
        for i in reversed(range(len(self.derivatives))):

            # get the last activation for the current layer
            activation = self.activations[i+1]

            # apply sigmoid derivative function
            delta = error * self._sigmoidPrime(activation)
            # save the derivative for gradient descent
            self.derivatives[i][:] = delta

            # backpropogate the next error
            error = np.dot(delta, self.weights[i].T)

            # save the error
            self.errors[i][:] = error

        return error

    def gradientDescent(self, learningRate=1):
        """Learns by descending the gradient

        Args:
            learningRate (float): How fast to learn.
        """
        # update the weights by stepping down the gradient
        for i in range(len(self.weights)):
            weights = self.weights[i]
            activations = self.activations[i+1]
            derivatives = self.derivatives[i]
            weights -= np.dot(activations, derivatives) * learningRate

    @staticmethod
    def _sigmoid(x):
        """Sigmoid activation function

        Args:
            x (float): Value to be processed

        Returns:
            y (float): Output
        """
        y = 1.0 / (1 + np.exp(-x))
        return y

    @staticmethod
    def _sigmoidPrime(x):
        return x * (1.0 - x)
